frontend/index.html

```diff
@@ -3,24 +3,46 @@
   <head>
     <meta charset="UTF-8" />
     <meta name="viewport" content="width=device-width, initial-scale=1.0" />
-    <!-- TAG INJECTION PLACEHOLDER -->
+    <title>{{ config.ui.name }}</title>
+    <link rel="icon" href="{{ favicon_path }}" />
+    <meta name="description" content="{{ config.ui.description }}">
+    <meta property="og:type" content="website">
+    <meta property="og:title" content="{{ config.ui.name }}">
+    <meta property="og:description" content="{{ config.ui.description }}">
+    <meta property="og:image" content="{{ meta_image_url }}">
+    <meta property="og:url" content="{{ url }}">
+    <meta property="og:root_path" content="{{ ROOT_PATH }}">
+
+    <!-- Preconnect for performance -->
     <link rel="preconnect" href="https://fonts.googleapis.com" />
     <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
-    <!-- FONT START -->
-    <link
-      href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;700&display=swap"
-      rel="stylesheet"
-    />
-    <!-- FONT END -->
-    <link
-      rel="stylesheet"
-      href="https://cdn.jsdelivr.net/npm/katex@0.16.8/dist/katex.min.css"
-    />
-    <!-- JS INJECTION PLACEHOLDER -->
-    <!-- CSS INJECTION PLACEHOLDER -->
-    <script>
-      const global = globalThis;
-    </script>
+
+    <!-- Conditionally include custom font or default Inter font -->
+    {% if config.ui.custom_font %}
+      <link rel="stylesheet" href="{{ config.ui.custom_font }}">
+    {% else %}
+      <link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;700&display=swap" rel="stylesheet" />
+    {% endif %}
+
+    <!-- Include KaTeX CSS if needed -->
+    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.8/dist/katex.min.css" />
+
+    <!-- Conditionally include custom CSS -->
+    {% if config.ui.custom_css %}
+      <link rel="stylesheet" type="text/css" href="{{ config.ui.custom_css }}">
+    {% endif %}
+
+    <!-- Theme Configuration Script -->
+    {% if config.ui.theme %}
+      <script id="theme-data" type="application/json">
+        window.theme = {{ config.ui.theme.to_dict() | tojson | safe }};
+      </script>
+    {% endif %}
+
+    <!-- Conditionally include custom JS -->
+    {% if config.ui.custom_js %}
+      <script src="{{ config.ui.custom_js }}" defer></script>
+    {% endif %}
```

Comment:
I find it's usually best to pass specific variables rather than more broad stuff to templates.

Rather than the entire `config`, I would pass just the contents of `ui`, so you'd have `custom_js`. This makes for more consistent adding of future variables, regardless of where they come from (e.g. the template editor doesn't have to care about whether something is config or something else).

Consider that these templates would be edited by someone who knows HTML but nothing of code, as is often the case.